saged/datasets.py

```diff
@@ -231,3 +244,443 @@ def get_all_data(self) -> np.array:
+            samples_to_keep = []
+            studies_to_keep = set()
+            last_study_index = 0
+            shuffled_studies = random.sample(studies, len(studies))
+
+            for study in shuffled_studies:
+                if len(samples_to_keep) > train_fraction * total_samples:
+                    break
+                studies_to_keep.add(study)
+
+                # This is inefficient compared to adding samples for each study as the study is
+                # added, but
+                samples_to_keep = utils.get_samples_in_studies(samples,
+                                                               studies_to_keep,
+                                                               self.sample_to_study)
+                last_study_index += 1
+
+            train_studies = shuffled_studies[:last_study_index]
+            test_studies = shuffled_studies[last_study_index:]
+
+        train_samples = utils.get_samples_in_studies(samples,
+                                                     train_studies,
+                                                     self.sample_to_study)
+        test_samples = utils.get_samples_in_studies(samples,
+                                                    test_studies,
+                                                    self.sample_to_study)
+
+        train_expression = self.all_expression.loc[:, train_samples]
+        test_expression = self.all_expression.loc[:, test_samples]
+
+        train_dataset = RefineBioLabeledDataset(train_expression,
+                                                self.sample_to_label,
+                                                self.sample_to_study,
+                                                )
+        test_dataset = RefineBioLabeledDataset(test_expression,
+                                               self.sample_to_label,
+                                               self.sample_to_study,
+                                               )
+        return train_dataset, test_dataset
+
+    def subset_samples_for_label(self, fraction: float,
+                                 label: str,
+                                 seed: int = 42,
+                                 ) -> "LabeledDataset":
+        """
+        Limit the number of samples available for a single label.
+        For example, if you wanted to use only ten percent of the sepsis expression
+        samples across all studies, you would use this function.
+
+        Arguments
+        ---------
+        fraction: The fraction of the samples to keep
+        label: The category of data to apply this subset to
+        seed: The seed for the random number generator involved in subsetting
+
+        Returns
+        -------
+        self: The subsetted version of the dataset
+        """
+        self.data_changed = True
+        random.seed(seed)
+
+        all_samples = self.get_samples()
+        samples_without_label = [sample for sample in all_samples
```

Comment:
If you want, you could probably partition `all_samples` in one pass (see, for example: https://stackoverflow.com/a/8793838).

That might be slightly less readable/understandable though (and not immediately clear to me whether it would actually be faster), so it might be best to stick with what you have if you don't expect this to be slow.

Reply:
Or something super simple like this is also an option: https://stackoverflow.com/a/58029106

Reply:
Ooh good idea